### PR TITLE
[X86] Extend kCFI with a 3-bit arity indicator

### PR DESCRIPTION
**Pull request by scottconstable**
Kernel Control Flow Integrity (kCFI) is a feature that hardens indirect calls by comparing a 32-bit hash of the function pointer's type against a hash of the target function's type. If the hashes do not match, the kernel may panic (or log the hash check failure, depending on the kernel's configuration). These hashes are computed at compile time by applying the xxHash64 algorithm to each mangled canonical function (or function pointer) type, then truncating the result to 32 bits. This hash is written into each indirect-callable function header by encoding it as the 32-bit immediate operand to a `MOVri` instruction, e.g.:
```
__cfi_foo:
	nop
	nop
	nop
	nop
	nop
	nop
	nop
	nop
	nop
	nop
	nop
	movl	$199571451, %eax                # hash of foo's type = 0xBE537FB
foo:
        ...
```

This PR extends x86-based kCFI with a 3-bit arity indicator encoded in the `MOVri` instruction's register (reg) field as follows:

| Arity Indicator | Description | Encoding in reg field |
| --------------- | --------------- | --------------- |
| 0 | 0 parameters | EAX |
| 1 | 1 parameter in RDI | ECX |
| 2 | 2 parameters in RDI and RSI | EDX |
| 3 | 3 parameters in RDI, RSI, and RDX | EBX |
| 4 | 4 parameters in RDI, RSI, RDX, and RCX | ESP |
| 5 | 5 parameters in RDI, RSI, RDX, RCX, and R8 | EBP |
| 6 | 6 parameters in RDI, RSI, RDX, RCX, R8, and R9 | ESI |
| 7 | At least one parameter may be passed on the stack | EDI |

For example, if `foo` takes 3 register arguments and no stack arguments then the `MOVri` instruction in its kCFI header would instead be written as:
```
	movl	$199571451, %ebx                # hash of foo's type = 0xBE537FB
```

This PR will benefit other CFI approaches that build on kCFI, such as FineIBT. For example, this proposed enhancement to FineIBT must be able to infer (at kernel init time) which registers are live at an indirect call target: https://lkml.org/lkml/2024/9/27/982. If the arity bits are available in the kCFI function header, then this information is trivial to infer.

Note that there is another existing PR proposal that includes the 3-bit arity within the existing 32-bit immediate field, which introduces different security properties: https://github.com/llvm/llvm-project/pull/117121.

### CODE REVIEW

File: llvm/lib/Target/X86/X86AsmPrinter.cpp [hunk 1]
```diff
@@ -181,8 +181,26 @@ void X86AsmPrinter::emitKCFITypeId(const MachineFunction &MF) {
   // Embed the type hash in the X86::MOV32ri instruction to avoid special
   // casing object file parsers.
   EmitKCFITypePadding(MF);
+
+  Register MovReg = X86::EAX;
+  const auto &Triple = MF.getTarget().getTargetTriple();
+  if (Triple.isArch64Bit() && Triple.isOSLinux()) {
```
Comment (scottconstable) [inline llvm/lib/Target/X86/X86AsmPrinter.cpp:187]: Is this check necessary? I notice that all of the x86-based kCFI tests use the `x86_64-unknown-linux-gnu` triple. If 64-bit Linux is the only supported x86 subtarget then I think this check is redundant. But I'm not 100% certain.

File: clang/include/clang/Basic/Features.def [hunk 2]
```diff
@@ -254,6 +254,7 @@ FEATURE(is_trivially_constructible, LangOpts.CPlusPlus)
 FEATURE(is_trivially_copyable, LangOpts.CPlusPlus)
 FEATURE(is_union, LangOpts.CPlusPlus)
 FEATURE(kcfi, LangOpts.Sanitize.has(SanitizerKind::KCFI))
+FEATURE(kcfi_x86_arity, LangOpts.Sanitize.has(SanitizerKind::KCFI))
```
Comment (scottconstable) [inline clang/include/clang/Basic/Features.def:257]: Peter Zijlstra, one of the Linux FineIBT maintainers, suggested to me that `__has_feature()` might be the best interface for Kconfig to discover that clang supports this new kCFI arity extension. If this feature were to be added to rustc, then I think rustc would need a similar interface to expose the feature.

File: llvm/lib/Target/X86/X86AsmPrinter.cpp [hunk 3]
```diff
@@ -181,8 +181,26 @@ void X86AsmPrinter::emitKCFITypeId(const MachineFunction &MF) {
...
+    // Determine the function's arity (i.e., the number of arguments) at the ABI
+    // level by counting the number of parameters that are passed
+    // as registers, such as pointers and 64-bit (or smaller) integers. The
+    // Linux x86-64 ABI allows up to 6 parameters to be passed in GPRs.
+    // Additional parameters or parameters larger than 64 bits may be passed on
+    // the stack, in which case the arity is denoted as 7.
+    const unsigned ArityToRegMap[8] = {X86::EAX, X86::ECX, X86::EDX, X86::EBX,
+                                       X86::ESP, X86::EBP, X86::ESI, X86::EDI};
```
Comment (phoebewang) [inline llvm/lib/Target/X86/X86AsmPrinter.cpp:195]: Astonished ESP can be used to pass the hash.

**Comment by phoebewang** [comment]
And you cannot use ESI when it's used to pass arguments. Suggested registers:
| Arity Indicator | Description | Encoding in reg field |
| --------------- | --------------- | --------------- |
| 0 | 0 parameters | EDI |
| 1 | 1 parameter in RDI | ESI |
| 2 | 2 parameters in RDI and RSI | EDX |
| 3 | 3 parameters in RDI, RSI, and RDX | ECX |
| 4 | 4 parameters in RDI, RSI, RDX, and RCX | R8D |
| 5 | 5 parameters in RDI, RSI, RDX, RCX, and R8 | R9D |
| 6 | 6 parameters in RDI, RSI, RDX, RCX, R8, and R9 | R10D |
| 7 | At least one parameter may be passed on the stack | EAX |

File: llvm/lib/Target/X86/X86AsmPrinter.cpp (hunk 1 above)
Comment (phoebewang) [reply llvm/lib/Target/X86/X86AsmPrinter.cpp:187]: You can use assert to guarantee it won't be broken in the future.

File: llvm/lib/Target/X86/X86AsmPrinter.cpp (hunk 3 above)
Comment (scottconstable) [reply llvm/lib/Target/X86/X86AsmPrinter.cpp:195]: Hi @phoebewang, based on my understanding of kCFI, the `MOVri` instruction emitted by the compiler in each function header will never be executed. The purpose of this instruction is simply to insert the function's kCFI type ID into the header so that it can be verified at the call site before making the call. For example, this is how a kCFI-enforced indirect call looks when the kernel is configured with kCFI:
```
0000000000000010 <foo>:    # foo loads the target address into rax
  ...
  25:   mov    $0x70e5d6ba,%r10d     # 0x70e5d6ba is the kCFI type ID that must be matched at the target
  2b:   add    -0x4(%rax),%r10d         # -0x4(%rax) refers to the immediate operand in the `MOVri` instruction, which holds the kCFI type ID
  2f:   je     33 <foo+0x23>    # If the two type IDs match, then jump to the call; otherwise #UD
  31:   ud2
  33:   call   *%rax      # Note that this will jump directly to the target function, and will not execute the `MOVri`
```
Does this address your concern?

**Comment by scottconstable** [comment]
> And you cannot use ESI when it's used to pass arguments.

@phoebewang I think that my https://github.com/llvm/llvm-project/pull/121070#discussion_r1898074522 also applies to this concern.

File: llvm/lib/Target/X86/X86AsmPrinter.cpp (hunk 1 above)
Comment (scottconstable) [reply llvm/lib/Target/X86/X86AsmPrinter.cpp]: @phoebewang Good suggestion! I updated the PR to use an assert.

File: clang/include/clang/Basic/Features.def (hunk 2 above)
Comment (Darksonn) [reply clang/include/clang/Basic/Features.def:257]: I still think it would be best to support both hash types via flags to avoid limiting which compiler versions work together.

**Comment by MaskRay** [comment]
If hashing is changed, consider replacing xxhash64 with xxh3+_64bits

File: clang/include/clang/Basic/Features.def (hunk 2 above)
Comment (scottconstable) [reply clang/include/clang/Basic/Features.def:257]: @Darksonn This PR differs from https://github.com/llvm/llvm-project/pull/117121 in that this PR does not change the hash type. Instead, this PR uses the reg field in the `MOVri32` instruction to encode some additional meta data about the function's arity.

**Comment by scottconstable** [comment]
> If hashing is changed, consider replacing xxhash64 with xxh3+_64bits

@MaskRay This PR does not change the hashing scheme at all.

**Comment by maurer** [comment]
> > If hashing is changed, consider replacing xxhash64 with xxh3+_64bits
> 
> @MaskRay This PR does not change the hashing scheme at all.

I think their point is that even if you are not changing the hash scheme, you are proposing breaking compatibility of the identifier with existing code. Since we don't want to do this many times, if we are breaking compatibility with existing code, they would like to batch it with another breaking update so that it doesn't need to be done again.

(This isn't me reviewing this PR, just trying to clear up some confusion.)

**Comment by MaskRay** [comment]
> maurer: I think their point is that even if you are not changing the hash scheme, you are proposing breaking compatibility of the identifier with existing code. Since we don't want to do this many times, if we are breaking compatibility with existing code, they would like to batch it with another breaking update so that it doesn't need to be done again.
>
> (This isn't me reviewing this PR, just trying to clear up some confusion.)

This:) Thanks for the explanation.

After my pending lld/MachO change, kcfi will be the only user of the legacy `xxHash64`. I want to remove `xxHash64` from llvm-project.

**Comment by scottconstable** [comment]
> > maurer: I think their point is that even if you are not changing the hash scheme, you are proposing breaking compatibility of the identifier with existing code. Since we don't want to do this many times, if we are breaking compatibility with existing code, they would like to batch it with another breaking update so that it doesn't need to be done again.
> > (This isn't me reviewing this PR, just trying to clear up some confusion.)
> 
> This:) Thanks for the explanation.
> 
> After my pending lld/MachO change, kcfi will be the only user of the legacy `xxHash64`. I want to remove `xxHash64` from llvm-project.

Going back to @maurer 's comment about compatibility, I do not believe that this PR would break any compatibility with other compilers, or with other compiler versions. If a kernel is built with some of its functions having the 3-bit arity extension and the rest of its functions not having the 3-bit arity extension (e.g., because the user's clang compiler has the extension, and their rust compiler does not), then the kernel will behave the same as it would if all functions or no functions have the 3-bit arity extension. If the kernel is hot-patched with the register hardening scheme in https://lkml.org/lkml/2024/9/27/982, then only the functions that have the 3-bit arity extension will benefit from the register hardening scheme.

**Comment by sirmc** [comment]
I had a quick look (sorry for not getting to this earlier, was traveling around Vietnam for a few weeks). Looks good to me. Especially the trick of encoding the arity into the MOV32ri register seems very neat from a compatibility standpoint.

So if I understand this correctly, the only compatibility conflict this introduces is with the kernel's handling of rewriting the kCFI function prologue, which can be checked with `__has_feature`. Enabling this arity hardening in Rust, if desired, seems straight-forward to me.

My only concern for the current PR was that the change might be incompatible with the handling on the kernel side (i.e., that the kernel's rewrite code assumed EAX rather than any other value in the reg field), which could fail older kernel builds with a newer toolchain. But after a quick look it seems like the decoding doesn't create conflict (see https://github.com/torvalds/linux/blob/c4b9570cfb63501638db720f3bee9f6dfd044b82/arch/x86/kernel/alternative.c#L1112). And AFAIK there's no other users of kCFI besides the Linux kernel (but please correct me if I'm wrong). I do see some initial work on supporting kCFI in FreeBSD https://reviews.freebsd.org/D46193, but this also seems compatible to me too. 

So in short, all seems ok to me.

**Comment by samitolvanen** [comment]
> A gentle reminder to please review this PR.

Overall this looks fine to me, but I was hoping to get an answer to my question about how the Kconfig detection is actually going to be implemented: https://github.com/llvm/llvm-project/pull/121070#discussion_r1911688455

**Comment by samitolvanen** [comment]
> And AFAIK there's no other users of kCFI besides the Linux kernel (but please correct me if I'm wrong).

There are firmware projects that use KCFI, but AFAIK none of them run on x86. I'm also fairly certain that only Linux performs runtime patching based on the code sequence we emit, so I'm not particularly concerned about breaking other KCFI users with this change.

**Comment by scottconstable** [comment]
> My only concern for the current PR was that the change might be incompatible with the handling on the kernel side (i.e., that the kernel's rewrite code assumed EAX rather than any other value in the reg field), which could fail older kernel builds with a newer toolchain.

@sirmc I shared this concern with Peter Zijlstra (who authored the Linux patches that will use this PR) and here is his response: _"This is correct; decode_preamble_hash() currently assumes the 0xb8 opcode. This is of course trivial to fix. Also, I think we'll need a compiler flag to enable this option (see next point), and since old kernel on new compiler will not set this flag, no problem."_

> Overall this looks fine to me, but I was hoping to get an answer to my question about how the Kconfig detection is actually going to be implemented: [#121070 (comment)](https://github.com/llvm/llvm-project/pull/121070#discussion_r1911688455)

@samitolvanen I also shared this concern with Peter, and he said: _"Yeah, so Kconfig will have to detect the compiler option is available for both C and Rust, and only select to use it when all (selected) languages support it. So if Rust is not enabled, we don't care if it supports the feature, and we can use it if clang supports it. But if Rust is enabled, both compilers need the feature present."_

I just updated the PR with a new clang flag `-fsanitize-kcfi-arity` that enables the new behavior. @Darksonn @lvwr @maurer @rcvalle your review and feedback would be appreciated.

**Comment by Darksonn** [comment]
If we can control it with a flag, then that sounds good to me.

**Review by samitolvanen — APPROVED** [review]
Thank you for adding the command line flag. I agree that this should be sufficient to address compatibility concerns.

**Review by phoebewang — APPROVED** [review]
LGTM.

**Comment by scottconstable** [comment]
@lvwr @maurer @rcvalle A gentle reminder to please review this PR.

**Comment by phoebewang** [comment]
> @lvwr @maurer @rcvalle A gentle reminder to please review this PR.

@scottconstable You don't need explicit approvals from all reviewers. Let's wait for 24 hours and land it if no objections.
